rpcs3/Emu/Cell/lv2/sys_uart.cpp

```diff
@@ -1,35 +1,597 @@
+		evnt->color.white_x = 0xFFFF;
+		evnt->color.gamma = 100;
+
+		evnt->supported_ai = 0;
+		evnt->speaker_info = 0;
+		evnt->num_of_audio_block = 0;
+
+		vuart.write_resp(pkt->hdr.cid, PS3AV_STATUS_SUCCESS, evnt.get(), sizeof(ps3av_get_monitor_info_reply));
+		return true;
+	}
+};
+
+struct av_get_hw_conf_cmd : public ps3av_cmd
+{
+	u16 get_size(vuart_av_thread &vuart, const void *pkt_buf) override
+	{
+		static_cast<void>(vuart);
+		static_cast<void>(pkt_buf);
+		return 8;
+	}
+
+	bool execute(vuart_av_thread &vuart, const void *pkt_buf, u32 reply_max_size) override
+	{
+		auto pkt = static_cast<const ps3av_header *>(pkt_buf);
+
+		if (reply_max_size < sizeof(ps3av_get_hw_info_reply) + sizeof(ps3av_pkt_reply_hdr))
+		{
+			vuart.write_resp(pkt->cid, PS3AV_STATUS_BUFFER_OVERFLOW);
+			return false;
+		}
+
+		// TODO:
+		auto out = std::make_unique<ps3av_get_hw_info_reply>();
+		out->num_of_hdmi = 1;
+		out->num_of_avmulti = 0;
+		out->num_of_spdif = 1;
+		out->resv = 0;
+
+		vuart.write_resp(pkt->cid, PS3AV_STATUS_SUCCESS, out.get(), sizeof(ps3av_get_hw_info_reply));
+		return true;
+	}
+};
+
+struct generic_reply_cmd : public ps3av_cmd
+{
+	u16 get_size(vuart_av_thread &vuart, const void *pkt_buf) override
+	{
+		static_cast<void>(vuart);
+		static_cast<void>(pkt_buf);
+		return 0;
+	}
+
+	bool execute(vuart_av_thread &vuart, const void *pkt_buf, u32 reply_max_size) override
+	{
+		auto pkt = static_cast<const ps3av_header *>(pkt_buf);
+
+		vuart.write_resp(pkt->cid, PS3AV_STATUS_SUCCESS);
+		return true;
+	}
+};
+
+error_code sys_uart_initialize(ppu_thread &ppu)
+{
+	ppu.state += cpu_flag::wait;
+
+	sys_uart.trace("sys_uart_initialize()");
+
+	static shared_mutex m{};
+	std::lock_guard lock(m);
+
+	auto &vuart_thread = g_fxo->get<vuart_av>();
+
+	if (vuart_thread.init)
+	{
+		return CELL_EPERM;
+	}
+
+	vuart_thread.init = 1;
 
 	return CELL_OK;
 }
 
-error_code sys_uart_receive(vm::ptr<void> buffer, u64 size, u32 unk)
+error_code sys_uart_receive(ppu_thread &ppu, vm::ptr<void> buffer, u64 size, u32 mode)
 {
-	sys_uart.todo("sys_uart_receive(buffer=*0x%x, size=0x%llx, unk=0x%x)", buffer, size, unk);
+	sys_uart.trace("sys_uart_receive(buffer=*0x%x, size=0x%llx, mode=0x%x)", buffer, size, mode);
 
-	return CELL_OK;
+	if (!size)
+	{
+		return CELL_OK;
+	}
+
+	if ((mode & ~1UL) != 0)
+	{
+		return CELL_EINVAL;
+	}
+
+	auto &vuart_thread = g_fxo->get<vuart_av>();
+
+	if (!vuart_thread.init)
+	{
+		return CELL_ESRCH;
+	}
+
+	if (size > 0x20000U)
+	{
+		// kmalloc restriction
+		fmt::throw_exception("Buffer is too big");
+	}
+
+	if (mode > 1)
+	{
+		return CELL_EINVAL;
+	}
+
+	const std::unique_ptr<u8[]> data = std::make_unique<u8[]>(size);
+	u32 read_size = 0;
+
+	auto vuart_read = [&](u8 *buf, u32 buf_size) -> s32
+	{
+		const u32 ITER_SIZE = 4096;
+		std::unique_lock lock(vuart_thread.rx_mutex, std::defer_lock);
+
+		if (!lock.try_lock())
+		{
+			return CELL_EBUSY;
+		}
+
+		const u32 first_iter_size = std::min(ITER_SIZE, buf_size);
```

Comment:
This section also has too much going on. As far as I can see it can be simplified to not have so many read_rx_data calls. Its basically (pseudocode):
```
u32 read_size;
for (auto remaining = buf_size, read_size= 0u; read_size< buf_size; )
{
    auto packet_size = std::min(remaining, ITER_SIZE);
    auto nread = read_rx_data(buf + read_size, packet_size);
    read_size+= nread;
    if (nread< packet_size) return position;
}
return read_size;
```

Reply:
Negative RE effects...

Reply:
Is it fixed? I looked over the new code and I don't see this pattern anymore.

Reply:
Lines 176-190